modules/compiler/test/group_ops.cpp

```diff
@@ -0,0 +1,402 @@
+
+class GroupOpsTest : public CompilerLLVMModuleTest {
+ public:
+  std::unique_ptr<PassMachinery> PassMach;
+
+  void SetUp() override {
+    CompilerLLVMModuleTest::SetUp();
+
+    auto Callback = [](const llvm::Module &) {
+      return compiler::utils::BuiltinInfo(
+          compiler::utils::createCLBuiltinInfo(/*builtins*/ nullptr));
+    };
+
+    PassMach = std::make_unique<compiler::BaseModulePassMachinery>(
+        Context, /*TM*/ nullptr, /*Info*/ multi_llvm::None, Callback,
+        /*verify*/ false, /*logging level*/ DebugLogging::None,
+        /*time passes*/ false);
+    PassMach->initializeStart();
+    PassMach->initializeFinish();
+  }
+
+  struct GroupOp {
+    GroupOp(StringRef FnName, StringRef LLVMTy, GroupCollective C)
+        : MangledFnName(FnName), LLVMTy(LLVMTy), Collective(C) {}
+
+    std::string getLLVMFnString(StringRef ParamName = "%x") const {
+      std::string FnStr =
+          LLVMTy + " @" + MangledFnName + "(" + LLVMTy + " " + ParamName.str();
+      if (Collective.Op == GroupCollective::OpKind::Broadcast) {
+        if (Collective.Scope == GroupCollective::ScopeKind::SubGroup) {
+          FnStr += ", i32 %sg_lid";
+        } else {
+          FnStr += ", i64 %lid_x, i64 %lid_y, i64 %lid_z";
+        }
+      }
+      FnStr += ")";
+      return FnStr;
+    }
+
+    std::string MangledFnName;
+    std::string LLVMTy;
+    GroupCollective Collective;
+  };
+
+  std::vector<GroupOp> getGroupBuiltinNames(GroupCollective::ScopeKind Scope,
+                                            bool IncludeAnyAll = true,
+                                            bool IncludeBroadcasts = true,
+                                            bool IncludeReductions = true,
+                                            bool IncludeScans = true) {
+    NameMangler Mangler(&Context);
+    Type *const I32Ty = IntegerType::getInt32Ty(Context);
+    Type *const I64Ty = IntegerType::getInt64Ty(Context);
+    Type *const FloatTy = IntegerType::getFloatTy(Context);
+
+    std::vector<GroupOp> GroupOpNames;
+    std::string BaseName =
+        std::string(Scope == GroupCollective::ScopeKind::SubGroup ? "sub"
+                    : Scope == GroupCollective::ScopeKind::VectorGroup
+                        ? "vec"
+                        : "work") +
+        "_group_";
+
+    GroupCollective Collective;
+    Collective.IsLogical = false;
+    Collective.Scope = Scope;
+
+    if (IncludeAnyAll) {
+      Collective.Op = GroupCollective::OpKind::Any;
+      Collective.Recurrence = RecurKind::Or;
+      GroupOpNames.emplace_back(GroupOp(
+          Mangler.mangleName(BaseName + "any", I32Ty, {eTypeQualSignedInt}),
+          "i32", Collective));
+
+      Collective.Op = GroupCollective::OpKind::All;
+      Collective.Recurrence = RecurKind::And;
+      GroupOpNames.emplace_back(GroupOp(
+          Mangler.mangleName(BaseName + "all", I32Ty, {eTypeQualSignedInt}),
+          "i32", Collective));
+    }
+
+    if (IncludeBroadcasts) {
```

Comment:
might be worth breaking this function up as it's quite long

Reply:
done